File: runtime/hetdesrun/persistence/models/operator.py
# noqa: A005
from uuid import UUID, uuid4

from pydantic import BaseModel, Field

from hetdesrun.models.code import NonEmptyValidStr, ShortNonEmptyValidStr
from hetdesrun.persistence.models.io import OperatorInput, OperatorOutput, Position
from hetdesrun.utils import State, Type


class Operator(BaseModel):
    """Represents components or workflows within a workflow.

    Contains all information about the transformation revision needed for display in the workflow as
    well as the id, so that the transformation itself can be loaded if more information is needed.

    Note: Only released transformation revisions can be used as operators in a workflow.
    """

    id: UUID = Field(default_factory=uuid4)  # noqa: A003
    revision_group_id: UUID = Field(default_factory=uuid4)
    name: NonEmptyValidStr
    type: Type  # noqa: A003
    state: State
    version_tag: ShortNonEmptyValidStr
    transformation_id: UUID
    inputs: list[OperatorInput]
    outputs: list[OperatorOutput]
    position: Position

File: frontend/src/app/model/operator.ts
import { TransformationType } from 'src/app/enums/transformation-type';
import { RevisionState } from 'src/app/enums/revision-state';
import { Connector } from './connector';
import { Position } from './position';

export interface Operator {
  /**
   * Id of enclosed operator (uuid)
   */
  id: string;

  /**
   * If operator belongs to a group, this is the groupId (uuid)
   */
  revision_group_id: string;
  name: string;

  /**
   * Type of operator (COMPONENT or WORKFLOW)
   */
  type: TransformationType;

  /**
   * State of operator (DRAFT, RELEASED, DISABLED)
   */
  state: RevisionState;
  version_tag: string;
  transformation_id: string;
  inputs: Connector[];
  outputs: Connector[];
  position: Position;
}
